### ISSUE TITLE
Add support & documentation for integrating with Lumen 5.2

### ISSUE THREAD

**Issue by rajabishek** [tymondesigns/jwt-auth#376]
Since Lumen does not support session state, incoming requests that we wish to authenticate must be authenticated via a stateless mechanism such as API tokens. This package is really going to be helpful for resolving out the authenticated user from the request through API token.
https://lumen.laravel.com/docs/5.2/authentication

``` php
Auth::viaRequest('api', function ($request) {
    try {
        if (! $user = JWTAuth::parseToken()->authenticate()) {
            return null;
        }
    } catch (Exception $e) {
        return null;
    } 
    return $user;
});
```

It would be really helpful if we have another section in the wiki about integrating this package with Lumen. I see that most of the work has already been done ( already there is a LumenServiceProvider ), only that it needs to be upgraded to support v5.2 and documented. Thank you Symon for giving us this wonderful package that we can use with out Laravel and Lumen projects.

**Comment by syropian** [comment]
@rajabishek Where do we put this `Auth::viaRequest` function? And what exactly is `api` here?

**Comment by syropian** [comment]
Weird, $user always return `null ` for me. It's the `null` from the `if` statement, not the `catch` though.

**Comment by rajabishek** [comment]
@syropian The AuthServiceProvider in Lumen 5.2 already has a call to Auth::viaRequest method that accepts a Closure which will be called when the incoming request needs to be authenticated ( if auth middleware is used ). Within this Closure, we may resolve out App\User instance however we wish. If no authenticated user can be found for the request, the Closure should return. So what I was suggesting was if we use this package with Lumen 5.2 then we would have to do something along the lines of 
```php
Auth::viaRequest('api', function ($request) {
    try {
        if (! $user = JWTAuth::parseToken()->authenticate()) {
            return null;
        }
    } catch (Exception $e) {
        return null;
    } 
    return $user;
});
```
And now once this is done, in our controllers or route handling closures we can directly get the authenticated user using the ```Auth::user()``` or through ```$request->user()``` just like how we would deal with sessions. Something like
```php
use Illuminate\Http\Request;

$app->get('/post/{id}', ['middleware' => 'auth', function (Request $request, $id) {
    $user = Auth::user();
    //or
    $user = $request->user();
}]);
```

So I just felt there must be a separate section in the wiki that needs to have this documented on how we can integrate and use this package with Lumen apps.

**Comment by syropian** [comment]
@rajabishek I tried setting mine up just like your example above, but `JWTAuth::parseToken()->authenticate()` is always null. I verified that the jwt token is present in the request but the authenticate method can never seem to fetch a user. Any ideas?

**Comment by rajabishek** [comment]
@syropian If you had used a similar code in the AuthServiceProvider that I had written above, then probably null might be returned because there is some exception beign thrown. Just remove the exception handler and probably we can get to see where the error is. Give this a try.

**Comment by syropian** [comment]
@rajabishek I actually tried that by just doing a really simple version like so:

```
Auth::viaRequest('api', function ($request) {
    $user = JWTAuth::parseToken()->authenticate();
    return $user;
});
```

but it's not returning a user (or a useful error). It's weird because I am able to generate the token for a given user no problem, but just can't get it to work the other way around! :c

**Comment by rajabishek** [comment]
@syropian Try doing this. 
```php
Auth::viaRequest('api', function ($request) {
    JWTAuth::parseToken();
    $token = JWTAuth::getToken();
    dd($token);
});
```
Just to run a check whether the package is actually able to grab the token from the request. If ```$token``` happens to be null, then we now know that there is some problem with the getting the token from the request headers, if so we can try setting the token explicitly by grabbing the Authorization header first and then the token value from that. Then we can set the token on the JWTAuth instance through the  ```JWTAuth::setToken($token);``` method. Before doing all this I would recommend to actually check once whether the request has the Authorization header. If there is no Authorization header then it could be a problem with Apache also. Reference: https://github.com/tymondesigns/jwt-auth/issues/200

**Comment by syropian** [comment]
@rajabishek Yup, I did check for that, the token is indeed present in the header. 

Anyway, I changed my code up a bit to ensure it returned null if the user wasn't present, and I get a 401 (unauthorized) response now. Very strange. Just to reiterate I have:

```
Auth::viaRequest('api', function ($request) {
  if( !$user = JWTAuth::parseToken()->authenticate() ){
    return null;
  }
  return $user;
});
```

and in my routes I have:

```
$app->post('document/create', ['middleware' => 'auth', function(Request $request, $id) {
    $user = Auth::user();
    dd($user);
}]);
```

I've made sure to enable the auth middleware in my `app.php` file.

**Comment by rajabishek** [comment]
@syropian You have the AuthServiceProvider service provider uncommented in your bootstrap/app.php file right ?

**Comment by syropian** [comment]
@rajabishek yup!

**Comment by syropian** [comment]
@rajabishek There isn't something internal that relies on the User model being in in the `App\User` namespace is there? I prefer to keep all my models in `App\Models\User`. I've updated the `jwt-auth` config to use this, and also updated the namespace in `AuthServiceProvider`.

**Comment by tdhsmith** [comment]
> I tried setting mine up just like your example above, but JWTAuth::parseToken()->authenticate() is always null.

Yeah check out my response in #384. You're using the stock Lumen setup, which creates an auth driver called `api`, which will be supported by `RequestGuard` (and the callback you pass to it in `viaRequest`). But the default jwt-auth setup has `authenticate()` calling `Auth::onceUsingId()`, which is not supported on `RequestGuard`, so you're going to get some sort of illegal method error.

There are a couple of solutions that I mention there:

1. Don't use `authenticate` (or its alias `toUser`), and define a viaRequest callback that works some other way (`$payload = JWTAuth::getPayload()` and then manually find the user from the `sub` claim with something like `App\Models\Whatever\User::find($payload['sub'])`).  If you go this route, you'll also have to find a workaround for `attempt`, if you use it.

2. Define a custom jwt-auth auth provider that doesn't rely on Laravel's Guards. The `byId` will basically be what I wrote in (1). The `byCredentials` depends on your needs, but will likely have to be capable of both finding a user by arbitrary credential queries AND validating the credentials (i.e. password hashing). Check out [`EloquentUserProvider`'s implementation of those functions](https://github.com/laravel/framework/blob/5.2/src/Illuminate/Auth/EloquentUserProvider.php#L87), for example.

3. Ultimately I think jwt-auth is going to be implemented as an auth driver for 5.2 and beyond, so instead of using the `AuthServiceProvider` at all, you'll just change your auth config to use driver `jwt`. You're more than welcome to help implement that change. :wink:

**Comment by tdhsmith** [comment]
Actually you might even be able to get by with changing the auth driver to `session` and then using the provided middleware as normal... jwt-auth doesn't actually require any of the session-y aspects of the driver, so I don't _think_ it would fail...?

**Comment by irazasyed** [comment]
Here's a package i created (Auth Driver) - https://github.com/irazasyed/jwt-auth-guard

For this package. Took me some time to play around with the code due to lot of various issues but at the end it was worth it.

Works like a charm.

P.S Obviously it would be good if the official package implements the Auth Driver.

**Comment by syropian** [comment]
@irazasyed That's awesome! Thanks a lot :)

@tdhsmith I ended up parsing the payload manually, and that worked just fine, thanks for the help :)

**Comment by irazasyed** [comment]
@syropian No problem, please leave your feedback/report any issues (if any) :)

**Comment by tymondesigns** [comment]
@irazasyed I've added the JWTGuard class now, although I don't have the time to test it properly right now. Feel free to give a spin guys :smile: 

And note that it's a first stab, so not *everything* is in there just yet

**Comment by jake142** [comment]
Could you please give some guidance on how to use the JWTGuard. Im trying to get jwt 0.6-dev to work with lumen 5.2 to but run into the 'Call to undefined method Illuminate\\Auth\\RequestGuard::once()'

I also have problem with the 'Class cache.store does not exist' in Tymon\JWTAuth\Providers\Storage\Illuminate (works if you use app('cache')).

Thank you for a great work.

**Comment by irazasyed** [comment]
@tymondesigns Looks good, I can see some errors just by looking at it though. Will try it out soon, Thanks :)

@jake142 You're having the same problems that i had reported before #384. There's some issue.
And the once undefined issue that you're seeing is because that method doesn't exist. @tymondesigns should probably implement some of those common methods, Like I've done in my package. I can send PR if that's ok.

**Comment by tdhsmith** [comment]
It might be easier to cut out the middleman and have `JWTGuard` implement `Tymon\JWTAuth\Contracts\Providers\Auth` directly? It doesn't really make a ton of sense for a stateless guard to implement `once` anyway, except to fulfill the requirements imposed by that intermediate provider.

At the very least there should probably be an auth provider specially for `JWTGuard`. That could be fixed right in the initial binding/setup.

**Comment by jake142** [comment]
@irazasyed Ah, ok, I thought that JWTGuard fixed the errors. Is the purpose of the JWTGuard to "replace" the default "guard"? If so, how do I configure my project to use the JWTGuard instead (sorry about the newbie questions)

**Comment by tdhsmith** [comment]
@jake142 Yes you are correct. The guard forms the basis of what becomes the `Auth` facade, and the center of the overall auth system. There are issues with Lumen because it has a very different default driver than Laravel does.

You would change it by setting in `config/auth.php`:
```php
'driver' => 'jwt'
```
In 5.2+, there will be an entry like that in the `guards` array for every corresponding label in the `defaults` array (though on stock Lumen there is only one default, called `'api'`).

Beware though, @tymondesigns just added this yesterday and it's mostly untested. There won't be much help available for problems you run into.

**Comment by jfoliveira** [comment]
Hi @tdhsmith 
I have this in my `config/auth.php`:
```php
<?php

return [
    'guards' => [
        'api' => [
            'driver' => 'jwt',
            'provider' => 'users',
        ],
    ],

    'providers' => [
        'users' => [
            'driver' => 'eloquent',
            'model' => App\Models\User::class,
        ],
    ],
];
```

And this is the error message I have when accessing an endpoint that requires authorization:
```
InvalidArgumentException in AuthManager.php line 71:
Auth guard [] is not defined.
```

Am I missing something in my `config/auth.php` file?

**Comment by naerymdan** [comment]
I added this so it knows that we the default guard is the api guard defined:

~~~
    'defaults' => [
        'guard' => 'api',
        'provider' => 'users'
    ],
~~~

But then in my case it fails at trying to instantiate AwtGuard from within JwtAuth, which causes a circular call, eventually crashing.

**Comment by jfoliveira** [comment]
@naerymdan thanks!
After adding the defaults I'm now facing the same issue as you: the circular reference is crashing my local Nginx. :-(

**Comment by naerymdan** [comment]
@jfoliveira yeah I've been trying to migrate a laravel+angular tutorial to Lumen, but a lot of dependencies have only partial support for Lumen, especially without Facades (why would anyone re-enable Facade, or worse, Eloquent on Lumen? Kinda defeats the purpose of using it in the first place).

https://github.com/naerymdan/lumen5-angular-material-starter <- if you are curious

**Comment by jfoliveira** [comment]
@irazasyed @tymondesigns any idea on why/where is the circular reference happening leading the application to crash?
I tried debugging the package for a few hours but it's still not clear to me what's the root cause of the issue.

**Comment by tymondesigns** [comment]
@jfoliveira I've just pushed an update which might solve the issue you mention. I wasn't happy with the implementation so far, so hopefully this will kill 2 birds with 1 stone. Let me know if the issue persists

**Comment by naerymdan** [comment]
@tymondesigns The newest code did resolve the circular reference on boot, but I'm getting it again when I try to use the JWTAuth instance in my controller:

~~~
class LoginController extends Controller
{
    protected $auth;

    public function __construct(JWTAuth $auth)
    {
        $this->auth = $auth;
    }

...
}
~~~

Mostly, it seems to circle around in the protected function extendAuthGuard() in LumenServiceProvider.php, when it tries to create the JwtGuard

**Comment by naerymdan** [comment]
@tymondesigns I was able to move forward by registering the JWT base class itself to use in the JwtGuard instantiation.

~~~
        $this->app->alias('tymon.jwt.auth.jwt', \Tymon\JWTAuth\JWT::class);
~~~

**Comment by jfoliveira** [comment]
@tymondesigns thanks for pushing this new update!

I've just updated and I am seeing this Fatal Error:
```
Fatal error: 
Cannot use Tymon\JWTAuth\Contracts\Providers\JWT as JWT because the name is already in use in
/vendor/tymon/jwt-auth/src/Manager.php on line 17
```

Changing `Manager.php` on line 17:
```php
// use Tymon\JWTAuth\Contracts\Providers\JWT;
use \Tymon\JWTAuth\JWT;
```

Leads to:
```
ErrorException in Manager.php line 50:
Argument 1 passed to Tymon\JWTAuth\Manager::__construct() must be an instance of 
Tymon\JWTAuth\JWT, 
instance of Tymon\JWTAuth\Providers\JWT\Namshi given, 
called in 
/vendor/tymon/jwt-auth/src/Providers/LumenServiceProvider.php on line 141
```

**Comment by tymondesigns** [comment]
@naerymdan @jfoliveira If you are using the Auth Guard feature in 5.2 then you don't need to use the `JWTAuth` facade at all. It's one or the other basically.

So just use the methods defined on the [Guard instance](https://github.com/tymondesigns/jwt-auth/blob/develop/src/JWTGuard.php) and you should be golden. e.g.

```php
$token = Auth::attempt($credentials); // get a token

Auth::user(); // get the authenticated user
Auth::setToken('foo.bar.baz')->user();

Auth::logout(); // invalidate the token a unset authenticated user
$refreshed = Auth::refresh(); // refresh the token parsed from the request

// etc...
```

**Comment by jfoliveira** [comment]
@tymondesigns thanks for commenting. I must be doing something really stupid then, as no matter what I do I'm getting PHP error:
```
Fatal error: Maximum function nesting level of '256' reached, aborting!
in /vendor/illuminate/container/Container.php on line 698
```

The route I'm using for testing is exactly that, with hardcoded credentials:

```php
$app->get('/auth', function (Request $request) {
    $credentials = ['email' => 'valid@email.com', 'password' => 'valid-password'];
    $token = Auth::attempt($credentials);
    dd($token);
});
```

In the routes file I'm just using `use Illuminate\Http\Request`. Should I reference `JWTAuth as Auth`?

My `config/auth.php` file:

```php
<?php

return [
    'defaults' => [
        'guard' => env('AUTH_GUARD', 'api'),
        'provider' => 'users',
    ],
    'guards' => [
        'api' => [
            'driver' => 'jwt',
        ],
    ],

    'providers' => [
        'users' => [
            'driver' => 'eloquent',
            'model' => App\Models\User::class,
        ],
    ],
];
```

My `bootstrap/app.php` file:
```php
$app->routeMiddleware([
    // 'auth' => App\Http\Middleware\Authenticate::class,
    'jwt.auth' => Tymon\JWTAuth\Middleware\Authenticate::class,
    'jwt.refresh' => Tymon\JWTAuth\Middleware\RefreshToken::class,
]);
...
$app->register(App\Providers\AppServiceProvider::class);
// $app->register(App\Providers\AuthServiceProvider::class);
$app->register(Tymon\JWTAuth\Providers\LumenServiceProvider::class);
```

Any ideas? :-(

**Comment by tymondesigns** [comment]
Whoops my bad. I forgot to update the service provider (it was late :smile:)

try running a `composer update`

also see #394 #368 #277

**Comment by jake142** [comment]
I'm getting 'Undefined index: provider' in LumenServiceProvider.

**Comment by naerymdan** [comment]
It works for me, but requires me to enable both Eloquent and Facades.

Without Facades I can't use the Auth::function() and simply injecting JWTGuard in my constructor fails  at building it.

Without Eloquent (using 'database' driver), it fails because GenericUser does not implement JWTSubject.

Is there any way to make this work without either Facades and Eloquent?

**Comment by jake142** [comment]
Ah, thanks. I just noticed that. Now I get: 
Argument 1 passed to Tymon\\JWTAuth\\JWT::fromUser() must be an instance of Tymon\\JWTAuth\\Contracts\\JWTSubject instance of App\\Models\\User given.

I use both facedes and eloquent. What am I doing wrong?

**Comment by naerymdan** [comment]
I personally added the interface and function to my User class, but there might be a better way:

~~~
class User extends Model implements
    AuthenticatableContract,
    AuthorizableContract,
    JWTSubject
{
...
    public function getJWTIdentifier() {
        return $this->id;
    }

    public function getJWTCustomClaims() {
        return [];
    }
}
~~~

**Comment by jake142** [comment]
Thank you, got it working now

**Comment by jfoliveira** [comment]
@tymondesigns working like a charm!
Thanks for your high quality support!

**Comment by jake142** [comment]
I use the middleware:

    'jwt.auth' => Tymon\JWTAuth\Middleware\Authenticate::class

If I change the user id in my database (seed a new database) and try to submit a valid token it passes through the middleware. It seems as if $this->auth->parseToken()->authenticate() returns false so that the unauthorized exception is never thrown:

    try {
        $this->auth->parseToken()->authenticate();
    } catch (JWTException $e) {
        throw new UnauthorizedHttpException('jwt-auth', $e->getMessage());
    }

I tested the Check middleware as well and got the same result. Am I doing something wrong?

**Comment by jake142** [comment]
I ran into 2 more problems.

When I do Auth::getPayload() I get 'A token is required' (JWTAuth::getPayload() works).

If I change the user ID after I issued the token (for example do a DB seed) the middleware Tymon\JWTAuth\Middleware\Authenticate and Tymon\JWTAuth\Middleware\Check doesn't throw an UnauthorizedHttpException but instead the $this->auth->parseToken()->authenticate() returns false

**Comment by tymondesigns** [comment]
@jake142 56f98aae78336e63bad3f2fa80279726a732f983b should have fixed that

**Comment by jfoliveira** [comment]
@tymondesigns I've just updated and `Auth::getPayload()` is now working. Thank you very much!

Another issue that might be related: calling `Auth::tokenById` as in:
```php
        if (!$token = Auth::tokenById($objUsuario->id)) {
            throw new UnauthorizedHttpException('Unauthorized.');
        }
```

Returns HTTP 400:
```json
{
  "message": "Token could not be parsed from the request."
}
```

Is a token needed to retrieve the token based on existing userId?
I also tried `Auth::fromUser` but it doesn't seem to be available on Lumen's `Auth` façade, so I guessed `tokenById` was the way to go.

**Comment by tdhsmith** [comment]
> Without Facades I can't use the `Auth::function()` and simply injecting `JWTGuard` in my constructor fails at building it.

@naerymdan What's failing in the constructor? Could you just use `app('auth')` or am I misunderstanding Lumen?

**Comment by naerymdan** [comment]
@tdhsmith Nope, you got it right, I was originally trying to inject JWTGuard directly in my constructor, which failed, but I succeeded in doing the following:

~~~
    /**
     * @var \Illuminate\Contracts\Auth\Factory
     */
    protected $auth;

    /**
     * @param  \Illuminate\Contracts\Auth\Factory $auth
     */
    public function __construct(Auth $auth)
    {
        $this->auth = $auth;
    }
~~~

And now i simply tried app('auth')... and it works -_-' I feel kinda... yeah.

**Comment by tdhsmith** [comment]
Haha no problem! I'm still not completely comfortable with the service container, so I frequently make silly mistakes with facades and singletons and injection and... :disappointed_relieved:

**Comment by jake142** [comment]
I'm not sure if this is an error or if Im doing something wrong. If I do the following:
1. Login my user and receive a token
2. Change the user id of that user (db seed for example)
3. Try to access a resource with the middleware Tymon\JWTAuth\Middleware\Authenticate::class

The filter doesn't throw an error (it passes through the middleware). I guess this is since the token is valid. If I remember correctly an error was thrown in version 0.5 (invalid token I think)?

**Comment by lukeed** [comment]
+10 for all the requests coming from #401 for Lumen 5.2 instructions.

If nothing else, getting everyone "up & running" the same way will help ensure new issues coming in are actual issues & not installation errors.

**Comment by BrettTrotter** [comment]
I started out with Lumen and 0.5 of your JWTAuth and had gotten everything working with some fiddling, but now that your project has evolved, I've struggled a bit to convert my project to use the new version of things. I don't suppose I could add an additional request specifically for documenting the conversion from 0.5 to present with Lumen in mind?

**Comment by Frandys** [comment]
i am try to get token using only Email but i am get error this one -> Argument 1 passed to Tymon\JWTAuth\JWT::fromUser() must be an instance of Tymon\JWTAuth\Contracts\JWTSubject, instance of App\Model\User given, called in C:\wamp\www\users-project\vendor\illuminate\support\Facades\Facade.php on line 217 and defined

i am using this code to Get token -> 
  $user=User::where('email','=','abc@gmail.com')->first();

if (!$userToken=JWTAuth::fromUser($user)) {
            return response()->json(['error' => 'invalid_credentials'], 401);
        }

return response()->json(compact('userToken'));

> Please help me how i get token using only rmail address
